src/psyclone/psyir/frontend/fparser2.py

```diff
@@ -1979,21 +1978,93 @@ def _process_derived_type_decln(self, parent, decl, visibility_map):
         # Populate this StructureType by processing the components of
         # the derived type
         try:
-            # We don't support derived-types with additional
-            # attributes e.g. "extends" or "abstract". Note, we do
+            # We support derived-types with additional attribute "extends"
+            # only.
+            # We do not support "abstract". Note, we do
             # support public/private attributes but these are stored
             # as Access_Spec, not Type_Attr_Spec.
             derived_type_stmt = decl.children[0]
-            if walk(derived_type_stmt, Fortran2003.Type_Attr_Spec):
-                raise NotImplementedError(
-                    "Derived-type definition contains unsupported attributes.")
-
-            # We don't yet support derived-type definitions with a CONTAINS
-            # section.
-            contains = walk(decl, Fortran2003.Contains_Stmt)
-            if contains:
-                raise NotImplementedError(
-                    "Derived-type definition has a CONTAINS statement.")
+            type_attr_spec_list = walk(derived_type_stmt,
+                                       Fortran2003.Type_Attr_Spec)
+            if type_attr_spec_list:
+                for type_attr_spec in type_attr_spec_list:
+                    # Deal with 'EXTENDS(parent_type)'.
+                    if type_attr_spec.items[0] == "EXTENDS":
+                        extends_name = type_attr_spec.items[1].string
+                        # Look up the extended type in the symbol table
+                        # and specialise the symbol if needed.
+                        if extends_name in parent.symbol_table:
```

Comment:
Thanks for spotting this, there's no reason for it to be in that precise table. 
Modified to use `lookup` as the parent` DataTypeSymbol` should not be added to the symbol table (or it will be defined, but it's often imported and then that's a redefinition, which will not compile). I added an explicit note about this.